lighthouse-core/report/html/renderer/report-ui-features.js

```diff
@@ -344,30 +347,31 @@ class ReportUIFeatures {
   /**
    * Opens a new tab to the online viewer and sends the local page's JSON results
    * to the online viewer using postMessage.
+   * @param {!ReportRenderer.ReportJSON} reportJson
+   * @param {string} viewerPath
    * @protected
    */
-  sendJsonReport() {
-    const VIEWER_ORIGIN = 'https://googlechrome.github.io';
-    const VIEWER_URL = `${VIEWER_ORIGIN}/lighthouse/viewer/`;
-
+  sendJsonReport(reportJson, viewerPath, onMessageCallback = function() { }) {
     // Chrome doesn't allow us to immediately postMessage to a popup right
     // after it's created. Normally, we could also listen for the popup window's
     // load event, however it is cross-domain and won't fire. Instead, listen
     // for a message from the target app saying "I'm open".
-    const json = this.json;
+    const json = reportJson;
     window.addEventListener('message', function msgHandler(/** @type {!Event} */ e) {
       const messageEvent = /** @type {!MessageEvent<{opened: boolean}>} */ (e);
       if (messageEvent.origin !== VIEWER_ORIGIN) {
         return;
       }
-
       if (messageEvent.data.opened) {
         popup.postMessage({lhresults: json}, VIEWER_ORIGIN);
-        window.removeEventListener('message', msgHandler);
```

Comment:
Is this because we need to listen for the render message to close? will this cause duplicate handler problems if they click "Open in viewer" twice?

Reply:
good question. @ebidel is that the situation this is handling?

Reply:
I remember one issue is if you open viewer from a report, then use viewer as a viewer (looking at other reports or something), if you refresh it will send back an `opened` message back to the original report and load the first report, not whatever you've been doing since.

Reply:
I think what @brendankenny described is the sitch. Although I'm not sure why Chrome does that, but you sometimes get another `message` event when the page refreshes.